File: src/ReLU.java
public class ReLU implements ActivationFunction{
    public Matrix apply(Matrix input) {
        Matrix output = input;
        for(int i =0;i<input.rows;i++)
            for (int j = 0 ;j < input.cols;j++)
                output.data[i][j] = Math.max(0,input.data[i][j]);
        return output;
    }

    public Matrix derivative(Matrix input) {
        Matrix output = input;
        for (int i =0;i<input.rows;i++)
            for(int j = 0 ;j < input.cols;j++)
                if(output.data[i][j] > 0){
                    output.data[i][j] = 1;
                }
            else
            output.data[i][j] = 0;
        return output;
    }
}

File: activation/ReLU.cpp
#include "ReLU.hpp"
#include <algorithm>

Matrix ReLU::activate(const Matrix& input) const {
    Matrix result(input.rows, input.cols);
    
    for (int i = 0; i < input.rows * input.cols; ++i) {
        result.data[i] = std::max(0.0, input.data[i]);
    }
    
    return result;
}

Matrix ReLU::derivative(const Matrix& input) const {
    Matrix result(input.rows, input.cols);
    
    for (int i = 0; i < input.rows * input.cols; ++i) {
        result.data[i] = (input.data[i] > 0.0) ? 1.0 : 0.0;
    }
    
    return result;
}

std::string ReLU::getName() const {
    return "ReLU";
}
